src/hotspot/share/cds/dumpTimeClassInfo.cpp

```diff
@@ -31,40 +31,58 @@
-  int clone_num_verifier_constraints = num_verifier_constraints();
-  if (clone_num_verifier_constraints > 0) {
-    clone._verifier_constraints = new (ResourceObj::C_HEAP, mtClass) GrowableArray<DTVerifierConstraint>(clone_num_verifier_constraints, mtClass);
-    clone._verifier_constraint_flags = new (ResourceObj::C_HEAP, mtClass) GrowableArray<char>(clone_num_verifier_constraints, mtClass);
-    for (int i = 0; i < clone_num_verifier_constraints; i++) {
-      clone._verifier_constraints->append(_verifier_constraints->at(i));
-      clone._verifier_constraint_flags->append(_verifier_constraint_flags->at(i));
+DumpTimeClassInfo::DumpTimeClassInfo(const DumpTimeClassInfo& src) {
+  assert(DynamicDumpSharedSpaces, "must be");
+
+  _klass = src._klass;
+  _nest_host = src._nest_host;
+  _failed_verification = src._failed_verification;
+  _is_archived_lambda_proxy = src._is_archived_lambda_proxy;
+  _has_checked_exclusion = src._has_checked_exclusion;
+  _id = src._id;
+  _clsfile_size = src._clsfile_size;
+  _clsfile_crc32 = src._clsfile_crc32;
+  _excluded = src._excluded;
+  _is_early_klass = src._is_early_klass;
+  _verifier_constraints = NULL;
+  _verifier_constraint_flags = NULL;
+  _loader_constraints = NULL;
+
+  assert(src._enum_klass_static_fields == NULL, "This should not happen with dynamic dump.");
+  _enum_klass_static_fields = NULL;
+
+  {
+    int n = src.num_verifier_constraints();
+    if (n > 0) {
+      _verifier_constraints = new (ResourceObj::C_HEAP, mtClass) GrowableArray<DTVerifierConstraint>(n, mtClass);
+      _verifier_constraint_flags = new (ResourceObj::C_HEAP, mtClass) GrowableArray<char>(n, mtClass);
+      for (int i = 0; i < n; i++) {
+        _verifier_constraints->append(src._verifier_constraints->at(i));
+        _verifier_constraint_flags->append(src._verifier_constraint_flags->at(i));
+      }
     }
   }
-  int clone_num_loader_constraints = num_loader_constraints();
-  if (clone_num_loader_constraints > 0) {
-    clone._loader_constraints = new (ResourceObj::C_HEAP, mtClass) GrowableArray<DTLoaderConstraint>(clone_num_loader_constraints, mtClass);
-    for (int i = 0; i < clone_num_loader_constraints; i++) {
-      clone._loader_constraints->append(_loader_constraints->at(i));
+
+  {
+    int n = src.num_loader_constraints();
+    if (n > 0) {
+      _loader_constraints = new (ResourceObj::C_HEAP, mtClass) GrowableArray<DTLoaderConstraint>(n, mtClass);
+      for (int i = 0; i < n; i++) {
+        _loader_constraints->append(src._loader_constraints->at(i));
+      }
     }
   }
-  assert(_enum_klass_static_fields == NULL, "This should not happen with jcmd VM.cds dumping");
-  return clone;
+}
+
+DumpTimeClassInfo::~DumpTimeClassInfo() {
+  if (_verifier_constraints != NULL) {
+    assert(_verifier_constraint_flags != NULL, "must be");
+    delete _verifier_constraints;
+    delete _verifier_constraint_flags;
+  }
+  if (_loader_constraints != NULL) {
+    delete _loader_constraints;
+  }
```

Comment:
So this delete operator in this destructor calls the destructors for all the elements of the _verifier_constraints, _verifier_constraint_flags and _loader_constraints elements, which decrements refcounts in them, right?

Reply:
Yes. `_loader_constraints` is a GrowableArray, which calls its element type's destructor.